ruby/ql/lib/codeql/ruby/frameworks/Pg.qll

```diff
@@ -0,0 +1,58 @@
+/**
+ * Provides modeling for Pg, a Ruby library (gem) for interacting with PostgreSQL databases.
+ */
+
+private import codeql.ruby.ApiGraphs
+private import codeql.ruby.dataflow.FlowSummary
+private import codeql.ruby.Concepts
+
+/**
+ * Provides modeling for Pg, a Ruby library (gem) for interacting with PostgreSQL databases.
+ */
+module Pg {
+  /**
+   * Flow summary for `PG.new()`. This method wraps a SQL string, marking it as
+   * safe.
+   */
+  private class SqlSummary extends SummarizedCallable {
+    SqlSummary() { this = "PG.new()" }
+
+    override MethodCall getACall() { result = any(PgConnection c).asExpr().getExpr() }
+
+    override predicate propagatesFlowExt(string input, string output, boolean preservesValue) {
+      input = "Argument[0]" and output = "ReturnValue" and preservesValue = false
+    }
+  }
+
+  /** A call to PG::Connection.open() is used to establish a connection to a PostgreSQL database. */
+  private class PgConnection extends DataFlow::CallNode {
+    PgConnection() {
+      this =
+        API::getTopLevelMember("PG")
+            .getMember("Connection")
+            .getAMethodCall(["open", "new", "connect_start"])
+      or
+      this = API::getTopLevelMember("PG").getAnInstantiation()
+    }
+  }
+
+  /** A call to `PG::Connection.open().exec`, considered as a SQL construction. */
+  private class PgConstruction extends SqlConstruction::Range, DataFlow::CallNode {
+    private DataFlow::Node query;
+
+    PgConstruction() {
+      exists(PgConnection pg_connection |
+        this =
+          pg_connection.getAMethodCall(["exec", "async_exec", "exec_params", "async_exec_params"]) and
+        query = this.getArgument(0)
+        or
+        this = pg_connection.getAMethodCall("prepare") and
+        this.getArgument(0).toString() =
+          pg_connection.getAMethodCall("exec_prepared").getArgument(0).toString() and
+        query = this.getArgument(1)
```

Comment:
If we change this to be an `SqlExecution` rather than an `SqlConstruction`, then we should probably have `this = pg_connection.getAMethodCall("exec_prepared")` rather than `this = pg_connection.getAMethodCall("prepare")`.

I also ran into some performance problems with this predicate relating to the join on `toString()` results. In general I wouldn't rely on comparing `toString()` results since they're more for human consumption than being reliable values to compare on. Using the corresponding `ConstantValue`s for these nodes seems to resolve the performance problem and should be more reliable. I'd rewrite this clause as something like:

```suggestion
        exists(string queryName, DataFlow::CallNode prepareCall |
          this = pgConnection.getAMethodCall("exec_prepared") and
          prepareCall = pgConnection.getAMethodCall("prepare") and
          prepareCall.getArgument(0).getConstantValue().isStringlikeValue(queryName) and
          this.getArgument(0).getConstantValue().isStringlikeValue(queryName) and
          query = prepareCall.getArgument(1)
        )
```